README.markdown
# Expense

Simple, personal expense tracking.

## Install

1. Clone the repository.
2. Copy the database.example.yml to database.yml and update if needed.
   <pre>
   cp config/database.example.yml config/database.yml
   </pre>
3. Ensure you have the required gems installed.
   <pre>
   rake gems:install
   </pre>
4. Create and migrate the database.
   <pre>
   rake db:create
   rake db:migrate
   </pre>

## iPhone

A simple iPhone-specific interface exists for expense entry and viewing.

![iPhone interface.](http://cloud.github.com/downloads/tristandunn/expense/expense.iphone.png)

## Daily builds

Code updates are regularly pushed to http://moneybag.heroku.com

## License

The MIT License

Copyright (c) 2009 Tristan Dunn

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG
CHANGELOG

16/03/2010
----------

- You can now add time to your inputs and it will get parsed and filtered
  accordingly

  "5.25 on subway for lunch 2 days ago"

  Full scale of the language processing can be found at the chronic rubydoc
  http://chronic.rubyforge.org/


spec/spec_helper.rb
# This file is copied to ~/spec when you run 'ruby script/generate rspec'
# from the project root directory.
ENV["RAILS_ENV"] ||= 'test'
require File.dirname(__FILE__) + "/../config/environment" unless defined?(RAILS_ROOT)
require 'spec/autorun'
require 'spec/rails'

Spec::Runner.configure do |config|
  # If you're not using ActiveRecord you should remove these
  # lines, delete config/database.yml and disable :active_record
  # in your config/boot.rb
  config.use_transactional_fixtures = true
  config.use_instantiated_fixtures  = false
  config.fixture_path = RAILS_ROOT + '/spec/fixtures/'

  # == Fixtures
  #
  # You can declare fixtures for each example_group like this:
  #   describe "...." do
  #     fixtures :table_a, :table_b
  #
  # Alternatively, if you prefer to declare them only once, you can
  # do so right here. Just uncomment the next line and replace the fixture
  # names with your fixtures.
  #
  # config.global_fixtures = :table_a, :table_b
  #
  # If you declare global fixtures, be aware that they will be declared
  # for all of your examples, even those that don't use them.
  #
  # You can also declare which fixtures to use (for example fixtures for test/fixtures):
  #
  # config.fixture_path = RAILS_ROOT + '/spec/fixtures/'
  #
  # == Mock Framework
  #
  # RSpec uses it's own mocking framework by default. If you prefer to
  # use mocha, flexmock or RR, uncomment the appropriate line:
  #
  # config.mock_with :mocha
  # config.mock_with :flexmock
  # config.mock_with :rr
  #
  # == Notes
  # 
  # For more information take a look at Spec::Runner::Configuration and Spec::Runner
end

require File.expand_path(File.dirname(__FILE__) + '/spec_helpers')


config/environments/test.rb
# Settings specified here will take precedence over those in config/environment.rb

# The test environment is used exclusively to run your application's
# test suite.  You never need to work with it otherwise.  Remember that
# your test database is "scratch space" for the test suite and is wiped
# and recreated between test runs.  Don't rely on the data there!
config.cache_classes = true

# Log error messages when you accidentally call methods on nil.
config.whiny_nils = true

# Show full error reports and disable caching
config.action_controller.consider_all_requests_local = true
config.action_controller.perform_caching             = false

# Disable request forgery protection in test environment
config.action_controller.allow_forgery_protection    = false

# Tell Action Mailer not to deliver emails to the real world.
# The :test delivery method accumulates sent emails in the
# ActionMailer::Base.deliveries array.
config.action_mailer.delivery_method = :test

# Use SQL instead of Active Record's schema dumper when creating the test database.
# This is necessary if your schema can't be completely dumped by the schema dumper,
# like if you have constraints or database-specific column types
# config.active_record.schema_format = :sql

config.gem 'rspec',       :lib => false, :version => '>= 1.2.6'
config.gem 'rspec-rails', :lib => false, :version => '>= 1.2.6'
config.gem 'timecop',                    :version => '>= 0.3.1'

config.time_zone = 'UTC'
config.after_initialize do
  Timecop.freeze(Time.gm(2009, 7, 1, 12))
end


spec/models/expense_spec.rb
require File.expand_path(File.dirname(__FILE__) + '/../spec_helper')

describe Expense do
  before do
    Time.zone = 'Central Time (US & Canada)'
  end

  describe 'class' do
    describe 'when calculating the sum of expenses for a given unit' do
      fixtures :expenses, :users

      it 'should return the correct totals' do
        t = Time.gm(2009,6,1,0,0)
        Timecop.freeze(t) do
          users(:bob).expenses.calculate_total_for(:month).should == 125.0
          users(:bob).expenses.calculate_total_for(:week).should == 25.0
          users(:bob).expenses.calculate_total_for(:day).should == 25.0
        end
      end
    end

    describe 'when calculating the average for a given unit' do
      before do
        Expense.stub!(:find_averages_for).and_return([5, 10])
        Expense.stub!(:determine_duration_since_first_entry_in).and_return(2)
      end

      it 'should return the average' do
        Expense.calculate_average_for(:day).should == 7.5
      end
    end

    describe 'when determing if the current average is above the overall average' do
      describe 'with records' do
        before do
          Expense.stub!(:first).and_return(true)
          Expense.stub!(:find_averages_for).and_return([4, 8])
          Expense.stub!(:calculate_average_for).and_return(2)
        end

        it 'should find the averages for the provided unit' do
          Expense.should_receive(:find_averages_for).with(:day)
          Expense.is_above_average_for?(:day)
        end

        it 'should calculate the averages for the provided unit' do
          Expense.should_receive(:calculate_average_for).with(:day)
          Expense.is_above_average_for?(:day)
        end

        describe 'when the current average is greater than the overall' do
          it 'should return true' do
            Expense.is_above_average_for?(:week).should == true
          end
        end

        describe 'when the current average is less than the overall' do
          before do
            Expense.stub!(:find_averages_for).and_return([2, 4])
            Expense.stub!(:calculate_average_for).and_return(4)
          end

          it 'should return false' do
            Expense.is_above_average_for?(:week).should == false
          end
        end
      end

      describe 'without records' do
        before do
          Expense.stub!(:first)
        end

        it 'should return false' do
          Expense.is_above_average_for?(:month).should == false
        end
      end
    end

    describe 'when finding recent expenses grouped by relative date' do
      fixtures :expenses, :users

      it 'should return an array of groups' do
        users(:default).expenses.find_recent_grouped_by_relative_date.to_a.should == [['Today', [expenses(:default)]]]
      end

      it 'should support a custom limit' do
        users(:default).expenses.find_recent_grouped_by_relative_date(0).to_a.should == []
      end
    end

    describe 'when finding averages by range' do
      fixtures :expenses, :users

      describe 'of days' do
        it 'should return averages for each day' do
          users(:bob).expenses.find_averages_for(:day).should == [7.50, 40.00, 60.00, 25.00]
        end
      end

      describe 'of weeks' do
        it 'should return averages for each week' do
          users(:bob).expenses.find_averages_for(:week).should == [7.50, 50.00, 25.00]
        end
      end

      describe 'of months' do
        it 'should return averages for each month' do
          users(:bob).expenses.find_averages_for(:month).should == [7.50, 41.67]
        end
      end
    end

    describe 'when determing the duration since the first entry' do
      describe 'with no entries' do
        before do
          Expense.stub!(:first).and_return(nil)
        end

        it 'should return one' do
          Expense.determine_duration_since_first_entry_in(nil).should == 1
        end
      end

      describe 'with at least one entry' do
        before do
          @first = mock('Expense')
          @first.stub!(:created_at).and_return(60.days.ago)

          Expense.stub!(:first).and_return(@first)
        end

        describe 'in days' do
          it 'should return number of days since first entry' do
            Expense.determine_duration_since_first_entry_in(:day).round.should == 60
          end
        end

        describe 'in weeks' do
          it 'should return number of weeks since first entry' do
            Expense.determine_duration_since_first_entry_in(:week).round.should == 9
          end
        end

        describe 'in months' do
          it 'should return number of months since first entry' do
            Expense.determine_duration_since_first_entry_in(:month).round.should == 2
          end
        end
      end
    end

    describe 'when determing the format for a range' do
      { :day => '%j%Y', :week => '%W%Y', :month => '%m%Y' }.each do |range, format|
        it "should return '#{format}' for #{range}" do
          Expense.determine_format_for(range).should == format
        end
      end
    end

    describe 'when searching' do
      fixtures :expenses

      it 'should find expenses by item' do
        Expense.search('cer').should       == [expenses(:bob_last_week)]
        Expense.search('grocer').should    == [expenses(:bob_last_week)]
        Expense.search('groceries').should == [expenses(:bob_last_week)]
      end

      describe 'and grouping by relative date' do
        it 'should find expenses by item and group by relative date' do
          Expense.search_grouped_by_relative_date('cer').to_a.should == [['Two Weeks Ago', [expenses(:bob_last_week)]]]
        end
      end
    end
  end

  describe 'when being created' do
    it 'should create valid expense' do
      lambda {
        create_expense
      }.should change(Expense, :count).by(1)
    end

    it 'should require user ID' do
      create_expense(:user_id => nil).errors.on(:user_id).should_not be_nil
    end

    it 'should require cost' do
      create_expense(:cost => nil).errors.on(:cost).should_not be_nil
    end

    it 'should require cost to be greater than zero' do
      create_expense(:cost =>  0).errors.on(:cost).should_not be_nil
      create_expense(:cost => -1).errors.on(:cost).should_not be_nil
    end

    it 'should require item' do
      create_expense(:item => nil).errors.on(:item).should_not be_nil
    end

    describe 'with no cost' do
      ['', 'on', 'for'].each do |separator|
        describe "and cost and item separated by '#{separator}" do
          before do
            @expense = create_expense(:cost => nil, :item => "5.45 #{separator} Subway for lunch")
          end

          it 'should extract cost from item' do
            @expense.cost.should == 5.45
          end

          it 'should remove cost from item' do
            @expense.item.should == 'Subway for lunch'
          end
        end
      end
    end
  end

  describe 'instance' do
    fixtures :expenses, :users

    before do
      @expense = expenses(:default)
    end

    it 'should belong to a user' do
      @expense.user.should == users(:default)
    end

    describe 'when determine relative date from today' do
      { 0    => 'Today',
        1    => 'Yesterday',
        2    => 'Last Week',
        7    => 'Two Weeks Ago',
        14   => 'Three Weeks Ago',
        21   => 'Four Weeks Ago',
        30   => 'Last Month',
        60   => 'Two Months Ago',
        90   => 'Three Months Ago',
        120  => 'Four Months Ago',
        140  => 'This Year',
        365  => 'Last Year',
        730  => 'Two Years Ago',
        1095 => 'Several Years Ago'
      }.each do |number_of, group|
        it "should return '#{group}' for #{number_of} days ago" do
          expense = create_expense(:created_at => number_of.days.ago)
          expense.relative_date.should == group
        end
      end
    end
  end

  protected

  def create_expense(options = {})
    returning(new_expense(options)) do |account|
      account.save
    end
  end

  def new_expense(options = {})
    Expense.new(valid_attributes.merge(options))
  end

  def valid_attributes
    { :user_id => 1,
      :cost    => 695.00,
      :item    => 'registration for RailsConf 2009.'
    }
  end
end


config/environment.rb
# Be sure to restart your server when you modify this file

# Specifies gem version of Rails to use when vendor/rails is not present
RAILS_GEM_VERSION = '2.3.5' unless defined? RAILS_GEM_VERSION

# Bootstrap the Rails environment, frameworks, and default configuration
require File.join(File.dirname(__FILE__), 'boot')

Rails::Initializer.run do |config|

  config.gem 'chronic'

  config.frameworks -= [ :active_resource, :action_mailer ]

  # Activate observers that should always be running
  # config.active_record.observers = :cacher, :garbage_collector, :forum_observer

  # Set Time.zone default to the specified zone and make Active Record auto-convert to this zone.
  # Run "rake -D time" for a list of tasks for finding time zone names.
  config.time_zone = 'Eastern Time (US & Canada)'

  # The default locale is :en and all translations from config/locales/*.rb,yml are auto loaded.
  # config.i18n.load_path += Dir[Rails.root.join('my', 'locales', '*.{rb,yml}')]
  # config.i18n.default_locale = :de
end


app/controllers/expenses_controller.rb
class ExpensesController < ApplicationController
# caches_action :index,
#               :cache_path => Proc.new { |controller|
#                 user   = controller.__send__(:current_user)
#                 format = controller.request.format.to_sym

#                 "index/user-#{user.id}.#{format}"
#               }

  # List recent expenses.
  def index
    load_expenses_and_averages

    @groups = @expenses.find_recent_grouped_by_relative_date
  end

  # Display new expense form.
  def new
    @expense = current_user.expenses.build
  end

  # Attempt to create an expense.
  #
  # If successful redirect, otherwise display the new
  # expense form.
  def create
    @expense = current_user.expenses.build(params[:expense])

    if @expense.save
      expire_fragment %r{index/user/#{current_user.id}\..*}

      redirect_to '/'
    else
      respond_to do |format|
        format.html   { render :action => :new }
        format.iphone { redirect_to '/' }
      end
    end
  end

  def destroy
    @expense = Expense.find(params[:id])
    @expense.destroy

    flash[:notice] = "Expenses destroyed"
    redirect_to "/"

  end

  # Search expenses.
  def search
    load_expenses_and_averages

    @query  = params[:search][:query]
    @groups = Expense.search_grouped_by_relative_date(@query)
  end

  protected

  # Load the generally required objects.
  def load_expenses_and_averages
    @expense  = current_user.expenses.build
    @expenses = current_user.expenses
    @averages = {
      :day   => @expenses.calculate_average_for(:day),
      :week  => @expenses.calculate_average_for(:week),
      :month => @expenses.calculate_average_for(:month)
    }
  end
end


app/controllers/application_controller.rb
# Filters added to this controller apply to all controllers in the application.
# Likewise, all the methods added will be available for all controllers.
class ApplicationController < ActionController::Base
  include Authentication

  # Include all helpers, all the time.
  helper :all

  # Require authentication by default.
  before_filter :login_required

  # Adjust the request form for iPhone requests.
  before_filter :adjust_format_for_iphone

  # See ActionController::RequestForgeryProtection for details.
  protect_from_forgery :only => [:create, :update, :destroy]

  # Scrub sensitive parameters from your log.
  filter_parameter_logging :password, :password_confirmation

  protected

  def adjust_format_for_iphone
    request.format = :iphone if iphone_request?
  end

  def iphone_request?
    request.env['HTTP_USER_AGENT'] &&
      request.env['HTTP_USER_AGENT'][/(AppleWebKit.+Mobile)/]
  end
end


app/models/expense.rb
class Expense < ActiveRecord::Base
  belongs_to :user

  before_validation :extract_cost_from_item, :extract_date_from_item

  validates_presence_of     :user_id
  validates_numericality_of :cost, :greater_than => 0
  validates_presence_of     :item

  # Determine the total for +unit+ of time.
  #
  # Valid options for +unit+ include:  month
  def self.calculate_total_for(unit)
    date = Time.now
    unit = unit.to_s
    case unit
      when "month"
        return self.sum(:cost, :conditions => {:created_at => (date.beginning_of_month..date.end_of_month)})
      when "week"
        return self.sum(:cost, :conditions => {:created_at => (date.beginning_of_week..date.end_of_week)})
      when "day"
        return self.sum(:cost, :conditions => {:created_at => (date.beginning_of_day..date.end_of_day)})
    end
  end

  # Determine the average for +unit+ of time.
  #
  # Valid options for +unit+ include: day, week, and month
  def self.calculate_average_for(unit)
    find_averages_for(unit).sum / determine_duration_since_first_entry_in(unit).to_f
  end

  # Determine if the latest average for +unit+ is above
  # the overall average.
  def self.is_above_average_for?(unit)
    return false unless first = first(:order => 'created_at ASC')

    find_averages_for(unit).first > calculate_average_for(unit)
  end

  # Find averages for a given +unit+ of time.
  def self.find_averages_for(unit)
    all(:order => 'created_at DESC').group_by do |expense|
      expense.created_at.strftime(determine_format_for(unit))
    end.collect do |group, expenses|
      (expenses.collect(&:cost).sum / expenses.length).round(2)
    end
  end

  # Find recent expenses, grouped by their relative date.
  def self.find_recent_grouped_by_relative_date(limit = 25)
    all(:order => 'created_at DESC', :limit => limit).group_by(&:relative_date)
  end

  # Search for expenses by a specific query.
  def self.search(query)
    all(:conditions => ['item LIKE ?', "%#{query}%"], :order => 'created_at DESC')
  end

  # Perform a search and group the results by relative date.
  def self.search_grouped_by_relative_date(query)
    search(query).group_by(&:relative_date)
  end

  # Determine the relative date from today.
  def relative_date
    case (Date.today - created_at.to_date)
    when 0         then 'Today'
    when 1         then 'Yesterday'
    when 2..6      then 'Last Week'
    when 7..13     then 'Two Weeks Ago'
    when 14..20    then 'Three Weeks Ago'
    when 21..29    then 'Four Weeks Ago'
    when 30..59    then 'Last Month'
    when 60..89    then 'Two Months Ago'
    when 90..119   then 'Three Months Ago'
    when 120..139  then 'Four Months Ago'
    when 140..364  then 'This Year'
    when 365..729  then 'Last Year'
    when 730..1094 then 'Two Years Ago'
    else                'Several Years Ago'
    end
  end

  protected

  # Determine the format for a +unit+ of time.
  def self.determine_format_for(unit)
    case unit
    when :day   then '%j%Y'
    when :week  then '%W%Y'
    when :month then '%m%Y'
    end
  end

  # Determine the duration in +unit+'s since the first entry.
  def self.determine_duration_since_first_entry_in(unit)
    return 1 unless first = first(:order => 'created_at ASC')

    duration = case unit
               when :day   then 1.day
               when :week  then 7.days
               when :month then 30.days
               end

    [1, (Time.now.to_f - first.created_at.to_f) / duration].max
  end

  # Extract the cost from an item, if none present.
  def extract_cost_from_item
    return unless cost.nil?

    if /^[$]?(\d+?\.\d{1,2}|\d+)\s*(on|for)*\s+(.*)$/.match(item)
      self.cost = $1
      self.item = $3
    end
  end

  def extract_date_from_item
      date = Chronic.parse(item)
      self.created_at = date
  end
end

